File: wildcardmatching/code.cpp
#include <iostream>
#include <unordered_set>

using namespace std;

class Solution
{
public:
    bool isMatch(string s, string p)
    {
        unordered_set<int> states;
        states.insert(0);

        for (char ch : p)
        {
            unordered_set<int> nextStates;
            for (int state : states)
            {
                if (ch == '*')
                {
                    for (int i = state; i <= s.size(); ++i)
                    {
                        nextStates.insert(i);
                    }
                }
                if (state >= s.size())
                    continue;
                if (s[state] == ch || ch == '?')
                {
                    nextStates.insert(state + 1);
                }
            }
            states = nextStates;
        }
        for (int state : states)
        {
            if (state == s.size())
                return true;
        }
        return false;
    }
};

int main()
{
    Solution solution;
    string s = "acdcb";
    string p = "a*c?b";
    bool result = solution.isMatch(s, p);
    cout << boolalpha << result << endl;
    return 0;
}

File: wildcardmatching/code.py
class Solution:
    def isMatch(self,s,p):
        states = {0}
        for ch in p:
            next_states = set()
            for state in states:
                if ch == '*':
                    for i in range(state, len(s) + 1):
                        next_states.add(i)
                if state >= len(s):
                    continue
                if s[state] == ch or ch == '?':
                    next_states.add(state + 1)
            states = next_states
        return len(s) in states

solution = Solution()
s = "acdcb"
p = "a*c?b"
result = solution.isMatch(s, p)
print(result)
